### PR TITLE
fix(updater): scan array source instead of pattern-matching it

### LINKED ISSUE #3099: A comment apostrophe inside `SYSTEM_PATHS` silently drops real paths from `extractArrayFromSource`
## Problem

`extractArrayFromSource` (`update-system.mjs:749`, v1.27.0) reads the path arrays by scraping every quoted run out of the array body:

```js
return Array.from(match[1].matchAll(/['"]([^'"]+)['"]/g), (entry) => entry[1]);
```

The regex has no concept of a comment, so a `'` or a `"` inside one is treated as a string delimiter. Two failure modes, both silent:

1. **An apostrophe consumes the surrounding entries.** The `'` in `upstream's` opens a string that closes on the opening quote of the next real path, so that path is destroyed and the comment text is returned in its place.
2. **A quoted phrase becomes a phantom path.** A comment like `// this bucket means "do not touch"` adds `do not touch` to the list as if it were a declared path.

Neither throws. The function returns a plausible-looking array that is simply wrong.

## Reproduction

```js
import { extractArrayFromSource } from './update-system.mjs';

const withApostrophe = `const SYSTEM_PATHS = [
  // upstream's own files
  'a.mjs',
  'b.mjs',
];`;

const withQuotes = `const SYSTEM_PATHS = [
  // this bucket means "do not touch"
  'c.mjs',
];`;

console.log(extractArrayFromSource(withApostrophe, 'SYSTEM_PATHS'));
console.log(extractArrayFromSource(withQuotes, 'SYSTEM_PATHS'));
```

**Expected:** `['a.mjs', 'b.mjs']` and `['c.mjs']`.

**Actual:**

```
[ "s own files\n  ", ",\n  " ]
[ "do not touch", "c.mjs" ]
```

In the apostrophe case *both* declared paths are gone, replaced by two fragments of the comment.

## Why it matters

The arrays are currently clean - I checked all 323 entries at v1.27.0 and every one is path-shaped, so nothing is broken today. It is the blast radius that makes it worth fixing rather than the current state:

- `update-system.mjs:1477` runs this against the **remote** updater source (`remoteSystemPaths = extractArrayFromSource(remoteUpdaterSource, 'SYSTEM_PATHS')`). A single apostrophe in a comment added upstream would therefore corrupt the path list on every client that upgrades, not just on the machine where it was typed.
- The result is the silent-wrong-answer shape #2919 describes: a dropped path means a real system file quietly stops being shipped, which is #2235 again, and a phantom path means the updater acts on an entry nobody declared.
- The trigger is ordinary English. `upstream's`, `the user's`, `it's`, or any quoted term in a comment is enough, and nothing in review would flag it as dangerous.

I hit it in a fork while declaring local-only paths, i.e. before #2421's `config/local-paths.txt` made that unnecessary. The declaration file removes the *user* path into this bug, but not the maintainer one.

## Proposed solution

Strip comments from the array body before scraping it:

```js
const body = match[1]
  .replace(/\/\*[\s\S]*?\*\//g, '')
  .replace(/\/\/[^\n]*/g, '');
return Array.from(body.matchAll(/['"]([^'"]+)['"]/g), (entry) => entry[1]);
```

Two smaller alternatives, if that feels like too much:

- Match only path-shaped strings, e.g. `/'([\w.@/-]+\/?)'/g`. Narrower, but it silently ignores a malformed entry instead of surfacing it.
- Validate after extraction and throw on anything that is not path-shaped. This does not fix the parse, but it converts a silent corruption into a loud failure, which is the more important half.

I am happy to open a PR with the comment-stripping version plus a test covering both shapes, since `CONTRIBUTING.md` says bug fixes can skip the issue step.

Would you rather have the parser fixed, or the post-extraction validation, or both?

### PR DESCRIPTION
**Pull request by IaroslavMazur**
Closes #3099.

## Problem

`extractArrayFromSource` finds the array with a non-greedy regex and scrapes string literals out of its body with a second one, and a regex cannot tell a quote in code from one in prose. Four shapes, all silent:

1. **An apostrophe in a comment eats its neighbours.** `// upstream's own files` opens a string that closes on the opening quote of the next declared path, destroying it and returning the comment text in its place.
2. **A quoted phrase in a comment adds a path nobody declared.** `// this bucket means "do not touch"` becomes an entry.
3. **A `];` in a comment ends the array early.** The non-greedy body match stops at the first occurrence in the text, not at the bracket that closes the array.
4. **A commented-out declaration wins.** The binding is matched wherever it appears, so `// const SYSTEM_PATHS = ['dead.mjs'];` is selected over the live one below it.

Nothing throws. The caller gets a plausible-looking list that is wrong.

## Why it is worth fixing rather than avoiding

Nothing is broken today - all 303 entries at v1.27.0 are path-shaped. It is the blast radius that matters:

1. **`apply` parses a source we did not write.** `remoteSystemPaths = extractArrayFromSource(remoteUpdaterSource, 'SYSTEM_PATHS')` reads the TARGET updater from `FETCH_HEAD`, so one apostrophe added upstream corrupts the manifest on every client that upgrades, not on the machine where it was typed.
2. **A dropped path is #2235 again.** A real system file quietly stops shipping, with no signal until something is missing.
3. **A phantom path hands the updater an entry no one wrote down.**
4. **The trigger is ordinary English.** `upstream's`, `the user's`, `it's`, or any quoted term in a comment is enough, and nothing in review reads as dangerous.

## The change

A single left-to-right scan replaces both regexes. Comments, string literals, escapes and regex literals are each consumed whole, and the array ends at the bracket that returns depth to zero rather than at the first `];` in the text.

The first revision of this PR stripped comments before scraping, which closed 1 and 2 but introduced a fifth shape, since the strip cannot see into a string either: `'https://host/file'` was truncated at the `//`, and its closing quote went with it, so the entry after it disappeared as well. Thanks to @coderabbitai for catching that one.

Two deliberate limits, both documented in the function:

- **Template-literal interpolation is read as ordinary text.** A path manifest has no reason to contain one, and reading it properly needs a parser rather than a scanner.
- **An unterminated array returns `[]`, not the entries collected so far.** `mergePathLists` merges an empty list away, so the caller falls back to its own manifest; it would act on a truncated one.

The one construct a character scan genuinely cannot resolve is `/` as division vs. the start of a regex literal. The standard preceding-token heuristic settles it, and the suite pins the case that matters here - a pattern containing an unbalanced quote sitting above the declaration.

## Verification

**Parity on real input.** `SYSTEM_PATHS`, `BOOTSTRAP_PATHS` and `USER_PATHS` extract identically under the old regex and the new scan - 303, 27 and 21 entries - against both `main` and this branch. The change is invisible to every array currently declared.

**Tests.** `tests/updater-extract-array-syntax.test.mjs` pins all four bug shapes, the URL regression, the regex-literal case, and three regression guards: a comment-free array must be unchanged, a missing binding must return `[]` rather than throw (a pre-1.11.0 target updater has no such array), and an unterminated one must report nothing.

5 of the 10 assertions fail against `main`, so the suite is not vacuous. The URL case is not among them - it passes on `main` and on this scan, and fails only under the intermediate comment strip, so it stands as a guard against reintroducing that approach.

**Suite.** 4426 passed / 0 failed on this branch, against 4416 passed / 0 failed on unmodified `main` - the difference is exactly these 10 assertions, with the same three pre-existing warnings.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved updater configuration parsing to correctly handle comments, quoted text, URLs, regular expressions, nested brackets, and escaped characters.
  * Prevented malformed or unterminated array declarations from producing incorrect entries.
  * Avoided extracting declarations that appear only inside comments.

* **Tests**
  * Added comprehensive coverage for valid, malformed, and edge-case updater configurations.

### CODE REVIEW

File: update-system.mjs [hunk 1]
```diff
@@ -747,10 +747,32 @@ function gitStatusEntries() {
...
+ * @returns {string[]} Declared entries, in source order. Empty when absent.
+ */
 export function extractArrayFromSource(source, name) {
   const match = source.match(new RegExp(`const\\s+${name}\\s*=\\s*\\[([\\s\\S]*?)\\];`));
   if (!match) return [];
-  return Array.from(match[1].matchAll(/['"]([^'"]+)['"]/g), (entry) => entry[1]);
+  const body = match[1]
+    .replace(/\/\*[\s\S]*?\*\//g, '')
+    .replace(/\/\/[^\n]*/g, '');
+  return Array.from(body.matchAll(/['"]([^'"]+)['"]/g), (entry) => entry[1]);
```
[the remark this answers is not in the document]
Comment (IaroslavMazur) [reply update-system.mjs]: > A declared entry such as `'https://host/file'` is truncated at `//`

Confirmed, and, unlike the other two, this one is a regression I introduced rather than a pre-existing gap:

```
main     : ['a.mjs', 'https://host/file', 'b.mjs']
bf7f4a38 : ['a.mjs', 'https:\n  ']
```

The strip eats the `//` inside the literal, the closing quote goes with it, and the next entry then pairs with the wrong delimiter - so one path is corrupted and its neighbour disappears. That is precisely the silent-corruption shape this PR exists to remove.

> a comment containing `];` can cause the match on Line 770 to stop before the real array end

Confirmed, and pre-existing; the strip changes its shape rather than introducing it:

```
main     : ['x']   // phantom entry scraped from inside the comment
bf7f4a38 : []      // body truncated at the comment, then emptied by the strip
```

On the `apply` path the blast radius is bounded, since `mergePathLists(SYSTEM_PATHS, remoteSystemPaths, BOOTSTRAP_PATHS)` falls back to the local manifest - so an empty remote list means new upstream files silently stop shipping (#2235 again) rather than nothing shipping at all.

> A commented-out `const SYSTEM_PATHS` can also be selected

Confirmed, pre-existing, and untouched by the strip: the binding is matched before anything is stripped, so both versions return `['dead.mjs']`.

> Track comments, quoted strings, escapes, and array boundaries with a scanner or parser instead.

Agreed, and done in 61d7e2fe. A single left-to-right scan replaces **_both_** regexes: comments, string literals, escapes and regex literals are each consumed whole, and the array now ends at the bracket that returns depth to zero rather than at the first `];` in the text. I kept it a scan rather than reaching for a parser, since the one construct a character scan genuinely cannot resolve is `/` as division vs. regex, and the standard preceding-token heuristic settles it - the suite pins that case with `/['"]/` sitting above the declaration.

Two things I deliberately did not do: template-literal interpolation is still read as ordinary text, documented as a known limit rather than handled, and an unterminated array now returns `[]` instead of the entries collected so far, on the grounds that `mergePathLists` merges an empty list away but would act on a truncated one.

Verification, since the whole point is that none of this throws:

1. **Parity on real input**: `SYSTEM_PATHS`, `BOOTSTRAP_PATHS` and `USER_PATHS` extract identically under the old regex and the new scan - 303, 27 and 21 entries - against both `main` and this branch.
2. **Non-vacuous tests**: 5 of the 10 assertions fail against `main`. The URL case is not one of them, since it only breaks under the intermediate strip, so it stands as a guard against reintroducing it.
3. **Suite**: 4426 passed, 0 failed, with the same three pre-existing warnings.

Title and description updated to match, since this is no longer about stripping comments.

Anything you would rather see handled differently?

**Comment by santifer** [comment]
Second one today and it lands in the most dangerous file in the repo, @IaroslavMazur. An apostrophe in a comment silently eating the next declared path is a failure nobody finds by reading, and a regex genuinely cannot tell a quote in code from one in prose — so replacing it with a real scan (string literals, regex-vs-division, character classes) is the honest fix rather than a smarter pattern. The detail that earns the merge is the failure mode: returning an empty list rather than a truncated one, because apply merges an empty list away but would act on a partial manifest. Merged 🚀
